docs/rfc/rfc-19-random-beacon-v2.adoc

```diff
@@ -0,0 +1,169 @@
+group creation was calculated based on the gas price ceiling, new groups were 
+getting created too often. 
+With too many groups being created, operators were spending too much ETH on
+ticket submission and too low rewards were getting accumulated per group making
+the reward withdrawal operation not profitable enough or not profitable at all
+in the case of small stakers.
+
+Extra submitter rewards added on top of reimbursements were hard to understand
+for staking providers and calculating the provider’s fee was complicated given
+the non-reimbursable cost of ticket submission. Surprisingly, it is much easier
+for staking providers to calculate their fee and costs in tBTC v1, having a
+clean rule that no operations are reimbursed and the only reward they receive
+is KEEP plus TBTC fees.
+
+=== Aggressive timeouts
+
+Relay entry timeouts were pretty aggressive. Set to 64x6 = 384 blocks, they were
+giving about 1.5 hours to submit a relay entry before the minimum stake from each
+group member was slashed. This timeout proved to be not enough to notify,
+investigate, and fix a problem. Three groups were slashed in a row during the
+November 11th incident that begun around 08:00 UTC, before any US-based
+operators noticed the problem.
+
+
+== Proposal
+
+[[fees-and-rewards]]
+=== Fees and rewards
+
+In tBTC v1, operators have to execute a lot of on-chain transactions per
+deposit. Even though they are not reimbursed for on-chain actions and they
+receive KEEP rewards and TBTC fees, this scheme proved to be successful and no
+operators complained about it.
+
+For v2 of the random beacon, all rewards and reimbursements should be done in T.
+This scheme is compatible with tBTC v2 design where all rewards and
+reimbursements are also done only in T.
+
+Relay requester should provide a fee in T. The value of the fee is a governable
+parameter. The fee is split into two parts. One part goes to a group creation
+pool and another part is used as a reimbursement for the relay entry submitter.
+What percent of the fee goes to the group creation pool is a governable
+parameter.
+
+Given that we want to incentivize the submission of a relay entry as soon as possible
+and that the price relation between gas price and T reimbursement for relay
+entry submission is unknown at the moment the fee is set by the governance, we
+should calculate a delay factor and delay penalty as in v1 of the random beacon.
+The delay penalty should be subtracted from reimbursement and sent to the
+sortition-pool-wide reward pool. The remaining part is sent to the relay entry
+submitter. Over the course of 64x6 blocks, reimbursement goes to zero. On-chain we
+are not going to enforce any order of submitters but off-chain clients should
+agree to the order where the member with index 1 has 6 blocks to submit, after 6
+blocks member with index 2 tries to submit, after 12 blocks member with index 3
+tries to submit and so forth.
```

Comment:
We can have this but someone is probably going to modify their client to front-run it whenever profitable; I definitely would.

Reply:
I think this is generally problematic if we do not enforce the order on-chain.

Options are:
- Implement a delay factor, do not enforce order on-chain - we can end up with the situation you described.
- Eliminate delay factor, do not enforce the order on-chain, the submitter is always getting the whole reward. This option is problematic because if the reward is hefty enough - and it probably has to be so that we do not end up adjusting it every day - we can end up with the same situation you described.
- Implement a delay factor, enforce order on-chain - last members in the group are in a much worse position. 
- Eliminate delay factor, enforce order on-chain - everyone in the group is responsible to submit the entry to avoid slashing, rules are the same for everyone.

It feels that eliminating delay factor and enforcing the order is the simplest option and most fair in the model when the entire sortition pool is rewarded. WDYT?

Reply:
Eliminating delay factor and enforcing the order has the problem that while any operator has an equal chance of being the first in the long term the windfall from having position 1 in a group would still be relatively significant. If we pay the full submitter reward but some operator has a persistent advantage in being the first one allowed to submit it would increase the noisiness of operator returns a lot.

If we use delay factor while enforcing order on-chain it would mean that we could start at zero reward and grow it until some operator determines it to be worth it to submit.

That would make entries slower to be submitted while eliminating the windfall from being the first; an actor controlling the first few operators could deliberately stall the submission to get a bit of an extra reward without risk of being frontrun by someone else, but that is unlikely to be massively impactful (someone controlling 25% of operators has a 1:1,000 chance of getting the first 5 members in a group).

Reply:
In the model proposed, we'd have:

delay factor = [ (T_deadline - T_received) / (T_deadline - T_begin)]^2
submitter reward = base reward * (1 - delay factor)

and there is no extra reward for other group members, correct?

The client would have to read KEEP/ETH price from some oracle and decide if it makes sense to submit. I like that the reward is adjusted to the current market conditions.

On the other hand, there are two pieces I dislike. First, it will take a longer time to submit a relay entry as you noted. Secondly, governance needs to adjust the base reward to make sure 100% of it covers the gas costs. It feels similar to the gas price ceiling. 

Another option would be not giving any submitter reward and assume rewards given to sortition pool members are enough. We would not enforce any order on-chain but off-chain, we'd implement a handshake agreement that the first member to submit is `entry % group_size`, then `entry % group_size + 1` and so on, and the only incentivize for submitting the entry is not to get slashed. It might not be that scary as it might feel given how ECDSA clients worked for tBTC v1. They were not reimbursed for any transaction and we had no complaints.

Reply:
> In the model proposed, we'd have:
> 
> delay factor = [ (T_deadline - T_received) / (T_deadline - T_begin)]^2
> submitter reward = base reward * (1 - delay factor)
> 
> and there is no extra reward for other group members, correct?

Yes. I had to read that math a couple of times but I think it's correct (reward starts at 0 and grows).

> On the other hand, there are two pieces I dislike. First, it will take a longer time to submit a relay entry as you noted. Secondly, governance needs to adjust the base reward to make sure 100% of it covers the gas costs. It feels similar to the gas price ceiling.

The base reward would not entirely need to cover the gas costs, as the group would get slashed if it misses the deadline; thus we only need to ensure that the amount slashed from a single operator covers the gas costs in all conditions.

If we want the price to increase faster, we could remove the `^2` and have the reward grow linearly. This increases the chance of the first submitter receiving a windfall but I think it's fine if the deadline is reasonably long.

> Another option would be not giving any submitter reward and assume rewards given to sortition pool members are enough. We would not enforce any order on-chain but off-chain, we'd implement a handshake agreement that the first member to submit is `entry % group_size`, then `entry % group_size + 1` and so on, and the only incentivize for submitting the entry is not to get slashed. It might not be that scary as it might feel given how ECDSA clients worked for tBTC v1. They were not reimbursed for any transaction and we had no complaints.

I like this way of ordering the submission order to spread the pole positions out over time, and this could even work with a windfall submitter reward that regularly exceeds gas costs. An off-chain handshake works reasonably well but it still has the problem that an operator/staking provider who makes their clients never submit entries (or only submit entries if they're rewarded above the gas cost) would still gain a benefit over time.

One possibility is that we could use the entry submission as a heartbeat; if you are in line for the submission (by the `entry % group_size` rule) but don't submit, you're deemed to be inactive and removed from the sortition pool, forcing the operator to rejoin the pool (possibly after a short ban period) and forfeit the rewards.

Reply:
I really like the idea of correlating the submission with a heartbeat. I think that if we eliminate most of the existing pricing complexity from the relay entry submission transaction, we could go below 200k gas (eyeballing) with just BLS verification and submission order check/inactivity. Assuming that the responsibility is spread between group members, and the gas cost is relatively low, would you be comfortable with eliminating the submitter reward for relay entry? I make a quiet assumption that sortition pool rewards are good enough so that operators are not complaining about 200k gas transaction submitted from time to time.

Reply:
> Assuming that the responsibility is spread between group members, and the gas cost is relatively low, would you be comfortable with eliminating the submitter reward for relay entry? I make a quiet assumption that sortition pool rewards are good enough so that operators are not complaining about 200k gas transaction submitted from time to time.

Yes, I think round robin duty to submit would be perfectly suitable without separate rewards.

If we want to penalise inactivity more than just being kicked out of the sortition pool, we could add slashing for each operator who failed to submit within their window. If the amount slashed per operator grows over time, and the number of slashed operators also grows, the total slashed amount would grow to the square of the time that has elapsed and it could also achieve the effect of bleeding the group if it is too late to submit.